js/src/TooltipContainer.js

```diff
@@ -20,53 +23,81 @@ TooltipContainer.prototype = {
+  this.tooltipRef = null
+
+  this.zoomContainer = zoomContainer
+  this.setup_zoom_callbacks(zoomContainer)
+
+  // Create callback manager
+  this.callback_manager = CallbackManager()
 
-  this.setup_zoom_callbacks(zoom_container)
+  this.div.on('mouseover', this.cancelHideTooltip.bind(this))
+  this.div.on('mouseleave', this.hide.bind(this))
 
-  // keep a reference to tinier tooltip
-  this.tooltip_component = tooltip_component
-  // if they pass in a function, then use that
-  this.tooltip_function = (_.isFunction(tooltip_component) ?
-                           function (state) { tooltip_component({ state: state, el: this.div.node() })} :
-                           null)
-  // if they pass in a tinier component, use that
-  this.tinier_tooltip = (tinier.checkType(tinier.COMPONENT, tooltip_component) ?
-                         tinier.run(tooltip_component, this.div.node()) :
-                         null)
+  // Attach a function to get size of tooltip
+  this.getTooltipSize = null
+  this.callback_manager.set('attachGetSize', getSizeFn => {
+    this.getTooltipSize = getSizeFn
+  })
 
+  this.map = null
   this.delay_hide_timeout = null
+  this.currentTooltip = null
+
+  // keep a reference to preact tooltip
+  preact.render(
+    <TooltipComponentContainer
+      callbackManager={this.callback_manager}
+      TooltipComponent={this.TooltipComponent}
+      tooltipRef={instance => { this.tooltipRef = instance }}
+      />,
+     this.div.node()
+  )
 }
 
 function setup_map_callbacks (map) {
-  this.placed_div = PlacedDiv(this.div, map)
+  this.map = map
+  this.placed_div = PlacedDiv(this.div, map, undefined, false)
 
   map.callback_manager.set('show_tooltip.tooltip_container', function (type, d) {
     if (map.settings.get_option('enable_tooltips')) {
       this.show(type, d)
     }
   }.bind(this))
-  map.callback_manager.set('hide_tooltip.tooltip_container',
-                           this.hide.bind(this))
-  map.callback_manager.set('delay_hide_tooltip.tooltip_container',
-                           this.delay_hide.bind(this))
+  map.callback_manager.set('hide_tooltip.tooltip_container', this.hide.bind(this))
+  map.sel.selectAll('#canvas').on('touchend', this.hide.bind(this))
+  map.callback_manager.set('delay_hide_tooltip.tooltip_container', this.delay_hide.bind(this))
+  map.callback_manager.set('update_tooltip.tooltip_container', function (type, sel) {
+    if (this.currentTooltip !== null) {
+      let d = null
+      sel.each(data => {
+        if (data[type.replace('_label', '_id')] === this.currentTooltip.id) {
+          d = data
+        }
+      })
+      if (d === null) {
+        throw Error(`Could not find tooltip data for ${this.currentTooltip}`)
+      }
+      this.show(type, d)
+    }
+  }.bind(this))
 }
 
-function setup_zoom_callbacks (zoom_container) {
-  zoom_container.callback_manager.set('zoom.tooltip_container', function () {
+function setup_zoom_callbacks (zoomContainer) {
```

Comment:
might be worth reverting to keep a clean diff. The large scale changes will happen in the `es6-classes` branch.

Reply:
Done